File: web/src/core/pipeline/CompressionPipeline.ts
import { CompressedContext } from '../model/CompressedContext.js';
import { CompressionLevel } from '../model/CompressionLevel.js';
import { NormalizedDocument } from '../model/NormalizedDocument.js';
import { AcceptanceCriteriaExtractionStage } from '../stage/AcceptanceCriteriaExtractionStage.js';
import { AssemblyStage } from '../stage/AssemblyStage.js';
import { BlockSplittingStage } from '../stage/BlockSplittingStage.js';
import { DecisionExtractionStage } from '../stage/DecisionExtractionStage.js';
import { DuplicateDetectionStage } from '../stage/DuplicateDetectionStage.js';
import { HtmlCleaningStage } from '../stage/HtmlCleaningStage.js';
import { NoiseFilterStage } from '../stage/NoiseFilterStage.js';
import { RankingStage } from '../stage/RankingStage.js';
import { RiskExtractionStage } from '../stage/RiskExtractionStage.js';
import { TodoExtractionStage } from '../stage/TodoExtractionStage.js';
import { PipelineStage } from './PipelineStage.js';
import { ProcessingContext } from './ProcessingContext.js';

/**
 * Orchestrates the deterministic compression stages. The default pipeline
 * needs no LLM; a semantic compressor can be layered on top of the CIR output.
 */
export class CompressionPipeline {
  private readonly stages: readonly PipelineStage[];

  constructor(stages: PipelineStage[]) {
    this.stages = [...stages];
  }

  static standard(): CompressionPipeline {
    return new CompressionPipeline([
      new HtmlCleaningStage(),
      new NoiseFilterStage(),
      new BlockSplittingStage(),
      new DuplicateDetectionStage(),
      new DecisionExtractionStage(),
      new AcceptanceCriteriaExtractionStage(),
      new TodoExtractionStage(),
      new RiskExtractionStage(),
      new RankingStage(),
      new AssemblyStage(),
    ]);
  }

  process(document: NormalizedDocument, level: CompressionLevel): CompressedContext {
    const context = new ProcessingContext(document, level);
    for (const stage of this.stages) {
      stage.process(context);
    }
    return context.result;
  }
}

File: crunchy-core/src/main/java/io/crunchy/core/pipeline/CompressionPipeline.java
package io.crunchy.core.pipeline;

import io.crunchy.core.model.CompressedContext;
import io.crunchy.core.model.CompressionLevel;
import io.crunchy.core.model.NormalizedDocument;
import io.crunchy.core.stage.AcceptanceCriteriaExtractionStage;
import io.crunchy.core.stage.AssemblyStage;
import io.crunchy.core.stage.BlockSplittingStage;
import io.crunchy.core.stage.DecisionExtractionStage;
import io.crunchy.core.stage.DuplicateDetectionStage;
import io.crunchy.core.stage.HtmlCleaningStage;
import io.crunchy.core.stage.NoiseFilterStage;
import io.crunchy.core.stage.RankingStage;
import io.crunchy.core.stage.RiskExtractionStage;
import io.crunchy.core.stage.TodoExtractionStage;

import java.util.List;

/**
 * Orchestrates the deterministic compression stages. The default pipeline
 * needs no LLM; a semantic compressor can be layered on top of the CIR output.
 */
public final class CompressionPipeline {

    private final List<PipelineStage> stages;

    public CompressionPipeline(List<PipelineStage> stages) {
        this.stages = List.copyOf(stages);
    }

    public static CompressionPipeline standard() {
        return new CompressionPipeline(List.of(
                new HtmlCleaningStage(),
                new NoiseFilterStage(),
                new BlockSplittingStage(),
                new DuplicateDetectionStage(),
                new DecisionExtractionStage(),
                new AcceptanceCriteriaExtractionStage(),
                new TodoExtractionStage(),
                new RiskExtractionStage(),
                new RankingStage(),
                new AssemblyStage()
        ));
    }

    public CompressedContext process(NormalizedDocument document, CompressionLevel level) {
        ProcessingContext context = new ProcessingContext(document, level);
        for (PipelineStage stage : stages) {
            stage.process(context);
        }
        return context.getResult();
    }
}
